### PR TITLE
PR-Audit-MINOR-Batch-2: parser strictness + slug length cap (blog_generation)

### PR DESCRIPTION
**Pull request by canfieldjuan**
**Plan:** [`plans/PR-Audit-MINOR-Batch-2.md`](../blob/claude/pr-audit-minor-batch-2/plans/PR-Audit-MINOR-Batch-2.md)

## Summary

Two audit MINORs in `blog_generation.py`. Several other audit findings I considered for this batch turned out to already be addressed by intermediate PRs (API status-code split, provider exception wrapping, scope sequence stripping all landed in later commits).

### Fix 1: `parse_blog_post_response` — relax to `if title:`

Pre-fix required both `title` AND `content` non-empty; missing-content drafts collapsed to `unparseable_response`. Now the parser identifies "is this a candidate" via non-empty `title`; the quality pack judges the rest (`content_too_short` fires on empty body). Same shape as the parser-strictness fix in PR-OptionA-1 for `parse_landing_page_response`.

### Fix 2: `_slugify` — cap at 100 chars

Pre-fix produced unbounded slugs (2000-char title → 2000-char slug). Capped at `_MAX_SLUG_CHARS = 100` with `rstrip("-")` to prevent trailing-hyphen artifacts after truncation.

## Intentional (looks wrong but is deliberate)

- **Parser keeps `title` as the minimum filter.** Could relax further (any well-formed JSON object), but that risks picking up unrelated objects in the response. Title is the cheapest "this is a blog candidate" signal.
- **100-char slug cap is the standard convention.** Not configurable per call — that would be plumbing without a real driver. If a host wants a different cap, they post-process before persistence.
- **Truncation rather than rejection.** A long-title draft is still valid; the slug just gets clipped.
- **`content` is still stripped on the parser path** (when present) so downstream consumers see normalized whitespace. Not gating, just normalizing.

## Deferred (still on purpose)

- `_accumulate_usage` cumulative-usage assumption (audit MINOR).
- `_landing_page_config_for_request` discards request (audit MINOR).
- Host concurrency assumption documentation (audit MINOR).
- `for_output` if/elif chain (audit NIT).
- `describe_control_surfaces` calls execution-services per request (audit MINOR) — needs separate caching design.
- `topic` for blog_post.
- `PR-Campaign-Config-V2`.

## Test plan

- [x] `pytest tests/test_extracted_blog_generation.py` → 15 passed
- [x] `python extracted/_shared/scripts/forbid_atlas_reasoning_imports.py extracted_content_pipeline` → clean
- [x] `bash scripts/check_ascii_python.sh` → passed

## Self-audit notes

Size: ~25 LOC source + ~50 LOC tests + ~95 LOC plan doc = ~170 LOC across 3 files. Comfortably under budget; the two fixes are small and the tests are tightly scoped.

https://claude.ai/code/session_013vD2xcxZG1cJEEwPYUvH97

---
_Generated by [Claude Code](https://claude.ai/code/session_013vD2xcxZG1cJEEwPYUvH97)_

### CODE REVIEW

File: extracted_content_pipeline/blog_generation.py [hunk 1]
```diff
@@ -71,8 +71,16 @@ def parse_blog_post_response(text: str) -> dict[str, Any] | None:
...
+        # ``title``); the quality pack judges the rest. Pre-fix required
+        # both ``title`` AND ``content`` so a missing-content draft would
+        # collapse to ``unparseable_response`` -- a missing-content blocker
+        # at the quality-pack layer is more useful. (Empty content fires
+        # ``content_too_short`` because zero words is below any min_words.)
+        # ``content`` is still stripped here when present so downstream
+        # consumers see normalized whitespace.
+        if title:
+            content = str(decoded.get("content") or "").strip()
             return {**dict(decoded), "title": title, "content": content}
```
Comment (canfieldjuan) [inline extracted_content_pipeline/blog_generation.py:84]: **Parser change: confirmed correct.**

`content = str(decoded.get("content") or "").strip()` returns `""` on a missing/null `content` key. Every downstream consumer (`_quality_check` at line ~326, `_build_draft` at line ~371) already calls `str(parsed.get("content") or "")` — safe on empty string, no crash risk.

The quality pack's word-count check in `blog_pack.py` does `word_count = len(body.split())` which is `0` for an empty string, and `0 < min_words` (default 1500) — so `content_too_short` fires unconditionally on empty body. The comment's claim is accurate.

File: extracted_content_pipeline/blog_generation.py [hunk 2]
```diff
@@ -393,9 +401,20 @@ def _blueprint_id(blueprint: Mapping[str, Any]) -> str:
...
 
 
+# PR-Audit-MINOR-Batch-2: cap slug length at 100 chars. Pre-fix, a
+# 2000-char title produced a 2000-char slug -- valid but unwieldy for
+# CMS / URL routing (most expect <100). Truncation rather than
+# rejection: a long-title draft is still valid; the slug just gets
+# clipped. ``rstrip("-")`` after the cut prevents trailing-hyphen
+# artifacts (``"a-very-long--"``).
+_MAX_SLUG_CHARS = 100
+
```
Comment (canfieldjuan) [inline extracted_content_pipeline/blog_generation.py:411]: **Slug cap: correct.**

`rstrip("-")` strips **all** trailing hyphens (Python's `str.rstrip` treats the argument as a character set, not a literal suffix), which is the desired behavior. The `slug or "blog-post"` fallback on the return line handles the degenerate case where truncation left only hyphens.

DB schema note: `blog_posts.slug` is `TEXT NOT NULL UNIQUE` (migration `084_blog_posts.sql`) with no length constraint, so 100 is a convention cap. The two independent `_slugify` helpers in the autonomous task files both cap at 60, creating a 100/60/60 split — not a problem, but worth noting for a future consolidation pass.

File: tests/test_extracted_blog_generation.py [hunk 3]
```diff
@@ -173,9 +173,17 @@ def test_parse_blog_post_response_strips_code_fences() -> None:
...
 def test_parse_blog_post_response_returns_none_when_required_fields_missing() -> None:
+    """PR-Audit-MINOR-Batch-2: parser only requires ``title`` as the
+    "is this a candidate" filter; ``content`` is delegated to the
+    quality pack (empty content fires ``content_too_short``)."""
     assert parse_blog_post_response("") is None
-    assert parse_blog_post_response('{"title": "No content"}') is None
+    # Missing title -> not a candidate.
     assert parse_blog_post_response('{"content": "No title"}') is None
+    # Missing content -> still a candidate; the quality pack judges it.
+    parsed = parse_blog_post_response('{"title": "No content"}')
```
Comment (canfieldjuan) [inline tests/test_extracted_blog_generation.py:183]: **Good test — covers the parser half of the claim.**

This confirms the new parser contract: `{"title": ..., no content}` → candidate returned with `content == ""`. The test would have failed on `main` where the old `if title and content:` guard caused this case to return `None`.

What's missing: an integration test through `generate()` confirming that this candidate produces `quality_blocked` with `content_too_short` blockers (rather than `unparseable_response`, the pre-fix behavior). See body for the suggested test shape.

File: tests/test_extracted_blog_generation.py [hunk 4]
```diff
@@ -396,3 +404,39 @@ async def test_generate_per_call_parse_retry_response_excerpt_chars_override():
...
     assert len(excerpt_section.rstrip()) <= 50
+
+
+# -----------------------
+# PR-Audit-MINOR-Batch-2: slug length cap + parser-strictness loosening
+# -----------------------
+
+
+def test_slugify_truncates_to_100_chars():
+    """A long title produces a clipped slug, not a 2000-char monstrosity."""
```
Comment (canfieldjuan) [inline tests/test_extracted_blog_generation.py:438]: **Smoking-gun slug test — correct.**

`long_title * 20` produces a slug well over 100 chars pre-fix, and exactly ≤100 post-fix. Would have failed on `main`.

One optional edge case: a title that slugifies to exactly 100 chars ending at a hyphen boundary. Given `rstrip("-")` after slicing and `slug or "blog-post"` as fallback, behavior is correct — adding the case is optional.

**Review by canfieldjuan — COMMENTED** [review]
## Review: PR-Audit-MINOR-Batch-2 (parser strictness + slug length cap)

Both fixes are **correct and safe to merge**. No blockers. Two findings worth addressing (one should-fix, one NIT) plus a consistency note.

---

### Summary

The parser relaxation is sound: every downstream consumer of `parsed["content"]` already guards with `str(parsed.get("content") or "")`, so an empty string passes through without crashing. The quality pack's word-count check fires `content_too_short` on an empty body (`len("".split()) == 0`, which is below `min_words` regardless of policy). The slug cap is correct, the constant is named, and `rstrip("-")` strips all trailing hyphens (Python semantics — set membership, not count). DB schema is `TEXT` with no length constraint, so 100 chars is a convention cap only, not a schema requirement.

The landing-page precedent claim is directionally right but not identical: `parse_landing_page_response` requires `title` AND a non-empty `sections` list as its minimum candidate filter, while the blog parser now requires only `title`. The spirit matches (minimum filter at parser, quality pack judges the rest), but reviewers looking for a direct copy of the pattern won't find it.

---

### Should-fix — Missing integration test for the new empty-content path

**`tests/test_extracted_blog_generation.py` (no diff line — describes a gap)**

The unit test at line 432 (`test_parse_blog_post_response_accepts_missing_content_for_quality_pack_to_judge`) verifies that the parser accepts a title-only JSON object and returns `content == ""`. That's correct. But the author's core claim is that the route change means *empty content now produces a `quality_blocked` error with `content_too_short` blockers, rather than `unparseable_response`*. That specific claim has no integration test. The existing `test_generate_blocks_low_quality_posts_without_saving` (line ~226) uses `content="Too short."` — it doesn't exercise the empty-content path this fix enables.

A smoking-gun test would be:
```python
@pytest.mark.asyncio
async def test_generate_routes_missing_content_to_quality_blocked_not_unparseable():
    """PR-Audit-MINOR-Batch-2: empty content -> quality_blocked, not unparseable_response."""
    service, _, blog_posts, _, _ = _service(
        responses=[_valid_blog_json(content="")],
        config=BlogPostGenerationConfig(
            quality_policy=QualityPolicy(
                name="blog_post",
                thresholds={"min_words": 20, "target_words": 20, "pass_score": 70},
            )
        ),
    )
    result = await service.generate(scope=TenantScope(), target_mode="vendor_retention", limit=1)
    assert result.generated == 0
    assert result.errors[0]["reason"] == "quality_blocked"  # not "unparseable_response"
    assert any("content_too_short" in b for b in result.errors[0]["blockers"])
    assert blog_posts.saved == []
```
This would have failed on `main` (empty content caused the parser to return `None` → `unparseable_response`) and passes post-fix, making it the definitive regression guard for the behavioral claim.

---

### NIT — Retry prompt still says "non-empty title and content"

**`extracted_content_pipeline/blog_generation.py` line 97 (out of diff)**

The retry user-prompt reads:
```python
"Return one JSON object with non-empty title and content. "
```
The parser now accepts title-only candidates, so the retry instruction is slightly inconsistent with the new contract. This doesn't cause any functional problem — telling the LLM to include both is still correct intent — but it's the one place in the file that contradicts the new rule. The analogous `_landing_page_user_prompt` says "non-empty title and sections" matching its parser's minimum requirements. Post-fix equivalent here would be: `"Return one JSON object with a non-empty title. "`. Low-risk NIT, fine to address in a follow-on.

---

### Consistency note (informational)

Three `_slugify` implementations now exist with different caps:
- `extracted_content_pipeline/blog_generation.py` — **100 chars** (this PR)
- `extracted_content_pipeline/autonomous/tasks/blog_post_generation.py:483` — **60 chars**
- `extracted_content_pipeline/autonomous/tasks/b2b_blog_post_generation.py:4945` — **60 chars**

All three slugs target the same `blog_posts.slug TEXT` column. Not a bug — just worth aligning on a future consolidation pass.

---

### Plan doc Fix 1 snippet discrepancy (informational)

The plan's "After" code block shows `return {**dict(decoded), "title": title}` omitting `content`. The actual implementation correctly includes `content`. Implementation is right; plan snippet is stale.

---

### What works well

- `_quality_check` and `_build_draft` both use `str(parsed.get("content") or "")` — empty string is safe throughout.
- `_MAX_SLUG_CHARS = 100` named constant is the right approach.
- `rstrip("-")` semantics are correct (strips all trailing hyphens, not just one).
- `slug or "blog-post"` fallback handles the degenerate case where truncation leaves only hyphens.
- Deferred items in the plan doc are verified real — `_accumulate_usage` and `_landing_page_config_for_request` are confirmed still open on `main`.

File: extracted_content_pipeline/blog_generation.py (hunk 1 above)
Comment (canfieldjuan) [inline extracted_content_pipeline/blog_generation.py:84]: **Parser change: confirmed correct.**

`content = str(decoded.get("content") or "").strip()` returns `""` on a missing/null `content` key. Every downstream consumer (`_quality_check` at line ~326, `_build_draft` at line ~371) already calls `str(parsed.get("content") or "")` — safe on empty string, no crash risk.

The quality pack's word-count check in `blog_pack.py` does `word_count = len(body.split())` which is `0` for an empty string, and `0 < min_words` (default 1500) — so `content_too_short` fires unconditionally on empty body. The comment's claim is accurate.

File: extracted_content_pipeline/blog_generation.py (hunk 2 above)
Comment (canfieldjuan) [inline extracted_content_pipeline/blog_generation.py:411]: **Slug cap: correct.**

`rstrip("-")` strips **all** trailing hyphens (Python's `str.rstrip` treats the argument as a character set, not a literal suffix), which is the desired behavior. The `slug or "blog-post"` fallback on the return line handles the degenerate case where truncation left only hyphens.

DB schema note: `blog_posts.slug` is `TEXT NOT NULL UNIQUE` (migration `084_blog_posts.sql`) with no length constraint, so 100 is a convention cap. The two independent `_slugify` helpers in the autonomous task files both cap at 60, creating a 100/60/60 split — not a problem, but worth noting for a future consolidation pass.

File: tests/test_extracted_blog_generation.py (hunk 3 above)
Comment (canfieldjuan) [inline tests/test_extracted_blog_generation.py:183]: **Good test — covers the parser half of the claim.**

This confirms the new parser contract: `{"title": ..., no content}` → candidate returned with `content == ""`. The test would have failed on `main` where the old `if title and content:` guard caused this case to return `None`.

What's missing: an integration test through `generate()` confirming that this candidate produces `quality_blocked` with `content_too_short` blockers (rather than `unparseable_response`, the pre-fix behavior). See body for the suggested test shape.

File: tests/test_extracted_blog_generation.py (hunk 4 above)
Comment (canfieldjuan) [inline tests/test_extracted_blog_generation.py:438]: **Smoking-gun slug test — correct.**

`long_title * 20` produces a slug well over 100 chars pre-fix, and exactly ≤100 post-fix. Would have failed on `main`.

One optional edge case: a title that slugifies to exactly 100 chars ending at a hyphen boundary. Given `rstrip("-")` after slicing and `slug or "blog-post"` as fallback, behavior is correct — adding the case is optional.

**Review by canfieldjuan — COMMENTED** [review]
## Review: PR-Audit-MINOR-Batch-2 (parser strictness + slug length cap)

Both fixes are **correct and safe to merge**. No blockers. Two findings worth addressing (one should-fix, one NIT) plus a consistency note.

---

### Summary

The parser relaxation is sound: every downstream consumer of `parsed["content"]` already guards with `str(parsed.get("content") or "")`, so an empty string passes through without crashing. The quality pack's word-count check fires `content_too_short` on an empty body (`len("".split()) == 0`, which is below `min_words` regardless of policy). The slug cap is correct, the constant is named, and `rstrip("-")` strips all trailing hyphens (Python semantics — set membership, not count). DB schema is `TEXT` with no length constraint, so 100 chars is a convention cap only, not a schema requirement.

The landing-page precedent claim is directionally right but not identical: `parse_landing_page_response` requires `title` AND a non-empty `sections` list as its minimum candidate filter, while the blog parser now requires only `title`. The spirit matches (minimum filter at parser, quality pack judges the rest), but reviewers looking for a direct copy of the pattern won't find it.

---

### Should-fix — Missing integration test for the new empty-content path

**`tests/test_extracted_blog_generation.py` (no diff line — describes a gap)**

The unit test at line 432 (`test_parse_blog_post_response_accepts_missing_content_for_quality_pack_to_judge`) verifies that the parser accepts a title-only JSON object and returns `content == ""`. That's correct. But the author's core claim is that the route change means *empty content now produces a `quality_blocked` error with `content_too_short` blockers, rather than `unparseable_response`*. That specific claim has no integration test. The existing `test_generate_blocks_low_quality_posts_without_saving` (line ~226) uses `content="Too short."` — it doesn't exercise the empty-content path this fix enables.

A smoking-gun test would be:
```python
@pytest.mark.asyncio
async def test_generate_routes_missing_content_to_quality_blocked_not_unparseable():
    """PR-Audit-MINOR-Batch-2: empty content -> quality_blocked, not unparseable_response."""
    service, _, blog_posts, _, _ = _service(
        responses=[_valid_blog_json(content="")],
        config=BlogPostGenerationConfig(
            quality_policy=QualityPolicy(
                name="blog_post",
                thresholds={"min_words": 20, "target_words": 20, "pass_score": 70},
            )
        ),
    )
    result = await service.generate(scope=TenantScope(), target_mode="vendor_retention", limit=1)
    assert result.generated == 0
    assert result.errors[0]["reason"] == "quality_blocked"  # not "unparseable_response"
    assert any("content_too_short" in b for b in result.errors[0]["blockers"])
    assert blog_posts.saved == []
```
This would have failed on `main` (empty content caused the parser to return `None` → `unparseable_response`) and passes post-fix, making it the definitive regression guard for the behavioral claim.

---

### NIT — Retry prompt still says "non-empty title and content"

**`extracted_content_pipeline/blog_generation.py` line 97 (out of diff)**

The retry user-prompt reads:
```python
"Return one JSON object with non-empty title and content. "
```
The parser now accepts title-only candidates, so the retry instruction is slightly inconsistent with the new contract. This doesn't cause any functional problem — telling the LLM to include both is still correct intent — but it's the one place in the file that contradicts the new rule. The analogous `_landing_page_user_prompt` says "non-empty title and sections" matching its parser's minimum requirements. Post-fix equivalent here would be: `"Return one JSON object with a non-empty title. "`. Low-risk NIT, fine to address in a follow-on.

---

### Consistency note (informational)

Three `_slugify` implementations now exist with different caps:
- `extracted_content_pipeline/blog_generation.py` — **100 chars** (this PR)
- `extracted_content_pipeline/autonomous/tasks/blog_post_generation.py:483` — **60 chars**
- `extracted_content_pipeline/autonomous/tasks/b2b_blog_post_generation.py:4945` — **60 chars**

All three slugs target the same `blog_posts.slug TEXT` column. Not a bug — just worth aligning on a future consolidation pass.

---

### Plan doc Fix 1 snippet discrepancy (informational)

The plan's "After" code block shows `return {**dict(decoded), "title": title}` omitting `content`. The actual implementation correctly includes `content`. Implementation is right; plan snippet is stale.

---

### What works well

- `_quality_check` and `_build_draft` both use `str(parsed.get("content") or "")` — empty string is safe throughout.
- `_MAX_SLUG_CHARS = 100` named constant is the right approach.
- `rstrip("-")` semantics are correct (strips all trailing hyphens, not just one).
- `slug or "blog-post"` fallback handles the degenerate case where truncation leaves only hyphens.
- Deferred items in the plan doc are verified real — `_accumulate_usage` and `_landing_page_config_for_request` are confirmed still open on `main`.
